### PR TITLE
Update ONNX Python tests

### PR DESCRIPTION
**Pull request by rblaczkowski**
* enable more ONNX python tests
	* increase XFAIL cases
	* small refactor

### CODE REVIEW

File: ngraph/python/tests/test_onnx/utils/__init__.py [hunk 1]
```diff
@@ -28,8 +28,43 @@
 from tests.test_onnx.utils.onnx_helpers import import_onnx_model
 
 
-def xfail_test(*backends, reason="Mark the test as expected to fail"):
-    return pytest.mark.xfail(condition=tests.BACKEND_NAME in backends, reason=reason, strict=True)
+def xfail_test(reason="Mark the test as expected to fail", strict=True):
+    return pytest.mark.xfail(reason=reason, strict=strict)
+
+
+segfault = pytest.mark.skip(reason="Segmentation fault error")
```
Comment (postrational) [inline ngraph/python/tests/test_onnx/utils/__init__.py:35]: Let's rename some od these decorators, to make it more obvious what they mean.
```suggestion
skip_segfault = pytest.mark.skip(reason="Segmentation fault error")
```

File: ngraph/python/tests/test_onnx/utils/__init__.py [hunk 2]
```diff
@@ -28,8 +28,43 @@
...
+issue_1 = xfail_test(reason="ValueError: could not broadcast input array")
```
Comment (postrational) [inline ngraph/python/tests/test_onnx/utils/__init__.py:36]:
```suggestion
xfail_reason_1 = xfail_test(reason="ValueError: could not broadcast input array")
```
and so on

File: ngraph/python/tests/test_onnx/utils/__init__.py [hunk 3]
```diff
@@ -28,8 +28,43 @@
...
+issue_16 = xfail_test(reason="RuntimeError: data [value] doesn't exist")
+issue_17 = xfail_test(reason="onnx.onnx_cpp2py_export.checker.ValidationError: "
+                             "Required attribute 'to' is missing.")
+issue_18 = xfail_test(reason="Assertion error - logsoftmax results mismatch")
+
+reduce_max = "ReduceMax"
+reduce_min = "ReduceMin"
+reduce_mean = "ReduceMean"
+reduce_sum = "ReduceSum"
+reduce_prod = "ReduceProd"
```
Comment (postrational) [inline ngraph/python/tests/test_onnx/utils/__init__.py:67]: These should be placed in `ngraph/python/tests/test_onnx/test_ops_reduction.py`

File: ngraph/python/tox.ini [hunk 4]
```diff
@@ -23,11 +23,8 @@ commands=
   flake8 {posargs:src/ setup.py}
   flake8 --ignore=D100,D101,D102,D103,D104,D105,D107,W503 tests/  # ignore lack of docs in tests
   mypy --config-file=tox.ini {posargs:src/}
-  ; TODO: uncomment the line below when all test are ready (and delete the following line)
-  ; pytest --backend={env:NGRAPH_BACKEND} {posargs:tests/}
-  pytest --backend={env:NGRAPH_BACKEND} tests/test_ngraph/test_core.py
-  pytest --backend={env:NGRAPH_BACKEND} tests/test_onnx/test_onnx_import.py tests/test_onnx/test_ops_binary.py
-
+  pytest --backend={env:NGRAPH_BACKEND} tests/test_ngraph/test_core.py -v
```
Comment (postrational) [inline ngraph/python/tox.ini:26]: Is the `test_ngraph` directory coming in another PR?

Comment (rblaczkowski) [reply ngraph/python/tox.ini:26]: yes

File: ngraph/python/tests/test_onnx/utils/__init__.py (hunk 1 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/utils/__init__.py]: done

File: ngraph/python/tests/test_onnx/utils/__init__.py (hunk 2 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/utils/__init__.py]: ok done, final schema is: xfail_issue_Number_in_Jira

File: ngraph/python/tests/test_onnx/utils/__init__.py (hunk 3 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/utils/__init__.py]: I prefer to have inside test suite files as many lines as is needed only for tests - but OK I moved.

File: ngraph/python/tests/test_onnx/test_ops_matmul.py [hunk 5]
```diff
@@ -18,7 +18,10 @@
 from onnx.helper import make_graph, make_model, make_node, make_tensor_value_info
 
 from tests.runtime import get_runtime
-from tests.test_onnx.utils import import_onnx_model
+from tests.test_onnx.utils import import_onnx_model, xfail_issue_35916, xfail_issue_35917, \
+    xfail_issue_35918, xfail_issue_35921
```
Comment (tsocha) [inline ngraph/python/tests/test_onnx/test_ops_matmul.py:22]: [SUGESTION]
Importing xfail_issues in the new line is more readable, isn't it?
```suggestion
from tests.test_onnx.utils import import_onnx_model, \
    xfail_issue_35916, \
    xfail_issue_35917, \
    xfail_issue_35918, \
    xfail_issue_35921
```

File: ngraph/python/tests/test_onnx/test_zoo_models.py [hunk 6]
```diff
@@ -559,7 +559,8 @@
         "url": _GITHUB_MODELS_LTS + "vision/classification/zfnet-512/model/zfnet512-8.tar.gz",
     },
     {
-        "model_name": "zfnet512_opset9",
+        "model_name": ""
+                      "",
```
Comment (tsocha) [inline ngraph/python/tests/test_onnx/test_zoo_models.py:563]: Is this change intentional?
Why did you change model name to an empty string?

File: ngraph/python/tox.ini [hunk 7]
```diff
@@ -23,11 +24,8 @@ commands=
...
+  pytest --backend={env:NGRAPH_BACKEND} tests/test_onnx -v -k 'not test_zoo_models.py'
```
Comment (tsocha) [inline ngraph/python/tox.ini:28]: Maybe we should just use this command here:
```suggestion
  pytest --backend={env:NGRAPH_BACKEND} {posargs:tests/}
```

File: ngraph/python/tests/test_onnx/test_zoo_models.py (hunk 6 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_zoo_models.py:563]: thanks, I missed to revert this change!

File: ngraph/python/tox.ini (hunk 7 above)
Comment (rblaczkowski) [reply ngraph/python/tox.ini:28]: tests directory consists of test_ngraph and test_onnx subdirectories. test_onnx is ready to use, test_ngraph not yet (will be pushed in another PR)

File: ngraph/python/tests/test_onnx/test_ops_matmul.py (hunk 5 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_matmul.py:22]: Thanks for suggestion, It is something with this approach. But anyway, we have a lot of tests files and one file with list of all xfail_issues. I think that one list is sufficient to understand how many issues we have. Also the names are so intuitive and all affected definition is wrapped also.

File: ngraph/python/tests/test_onnx/test_ops_matmul.py [hunk 8]
```diff
@@ -99,33 +102,20 @@ def import_and_compute_gemm(input_a, input_b, input_c, **kwargs):
...
+        (([1, 2, 3], [[4], [5], [6]]), "vector at vector 2"),
+        (([[1, 2, 3]], [1, 2, 3]), "vector at vector 3"),
+        (([1, 2, 3], [[4, 5], [6, 7], [8, 9]]), "vector at matrix"),
+        (([[1, 2, 3], [4, 5, 6]], [[7], [8], [9]]), "matrix at vector"),
+        (([[1, 2], [3, 4]], [[5, 6], [7, 8]]), "matrix at matrix 1"),
+        (([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10], [11, 12]]), "matrix at matrix 2"),
+        (([[1, 2], [3, 4], [5, 6]], [[7, 8, 9], [10, 11, 12]]), "matrix at matrix 3")
+    ],
+)
+def test_op_matmul(data, description):
```
Comment (arogowie-intel) [inline ngraph/python/tests/test_onnx/test_ops_matmul.py:118]: Great refactoring!

File: ngraph/python/tests/test_onnx/test_ops_matmul.py (hunk 5 above)
Comment (arogowie-intel) [reply ngraph/python/tests/test_onnx/test_ops_matmul.py:22]: If I'm not mistaken following syntax is also correct:
```
from tests.test_onnx.utils import (
  import_onnx_model,
  xfail_issue_35916,
 # and so on..
)

File: ngraph/python/tests/test_onnx/test_ops_nonlinear.py [hunk 9]
```diff
@@ -103,13 +104,19 @@ def selu(x, alpha=1.67326319217681884765625, gamma=1.05070102214813232421875):
...
-    def elu(x, alpha=1):
+    def elu(x, alpha):
         return np.where(x < 0, alpha * (np.exp(x) - 1), x)
 
-    assert_onnx_import_equals_callable("Elu", elu, [-2, -1.0, 0.0, 1.0, 2.0])
-    assert_onnx_import_equals_callable("Elu", elu, [0.0])
-    assert_onnx_import_equals_callable("Elu", elu, [-0.9, -0.8, -0.7, -0.4, -0.3, -0.2, -0.1])
-    assert_onnx_import_equals_callable("Elu", elu, [[1, 2, 3], [4, 5, 6]])
-    assert_onnx_import_equals_callable("Elu", elu, [-2, -1.0, 0.0, 1.0, 2.0], alpha=0.5)
+    assert_onnx_import_equals_callable("Elu", elu, [-2, -1.0, 0.0, 1.0, 2.0], alpha=alpha_value)
```
Comment (arogowie-intel) [inline ngraph/python/tests/test_onnx/test_ops_nonlinear.py:122]: I guess you want to use here test parameter: 
```suggestion
    assert_onnx_import_equals_callable("Elu", elu, data, alpha=alpha_value)
```

File: ngraph/python/tests/test_onnx/test_ops_reduction.py [hunk 10]
```diff
@@ -17,7 +17,34 @@
...
+
+reduce_max = "ReduceMax"
+reduce_min = "ReduceMin"
+reduce_mean = "ReduceMean"
+reduce_sum = "ReduceSum"
+reduce_prod = "ReduceProd"
+
+reduce_data = np.array([[[5, 1], [20, 2]], [[30, 1], [40, 2]], [[55, 1], [60, 2]]], dtype=np.float32)
+reduce_axis_parameters = [
+    # (operation, keepdims - True/False, axis)
```
Comment (arogowie-intel) [inline ngraph/python/tests/test_onnx/test_ops_reduction.py:30]: I'm not sure how this comment relates to the following code...

File: ngraph/python/tests/test_onnx/test_ops_reduction.py [hunk 11]
```diff
@@ -26,196 +53,17 @@ def import_and_compute(op_type, input_data, **node_attrs):
...
-        np.prod(data, keepdims=False, axis=(0, 1, 2)),
-    )
+@unstrict_xfail_issue_35925
+@pytest.mark.parametrize("operation", reduce_operation_parameters)
+@pytest.mark.parametrize("keepdims", [True, False])
+@pytest.mark.parametrize("axes", reduce_axis_parameters)
+def test_reduce_operation(operation, keepdims, axes):
+    if axes:
+        assert np.array_equal(import_and_compute(operation, reduce_data, axes=axes, keepdims=keepdims),
+                              np.max(reduce_data, keepdims=keepdims, axis=axes))
```
Comment (arogowie-intel) [inline ngraph/python/tests/test_onnx/test_ops_reduction.py:63]: You have to also parameterize this numpy reference function.

File: ngraph/python/tests/test_onnx/test_zoo_models.py (hunk 6 above)
Comment (arogowie-intel) [inline ngraph/python/tests/test_onnx/test_zoo_models.py:563]: I still see empty model name.

Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_zoo_models.py:563]: Yep, sorry, I commited  this change but didn't push to upstream

File: ngraph/python/tests/test_onnx/test_ops_matmul.py (hunk 5 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_matmul.py:22]: > If I'm not mistaken following syntax is also correct:
> 
> ```
> from tests.test_onnx.utils import (
>   import_onnx_model,
>   xfail_issue_35916,
>  # and so on..
> )
> ```

It doesn't work in Python. I propose then the approach:
import (lib1,
             lib2,
             ...,
             )

File: ngraph/python/tests/test_onnx/test_ops_nonlinear.py (hunk 9 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_nonlinear.py:122]: Oh, yes, thanks to point it out

File: ngraph/python/tests/test_onnx/test_ops_reduction.py (hunk 10 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_reduction.py:30]: Now this comment is deprecated - it was related to previous version  - fixed

File: ngraph/python/tests/test_onnx/test_ops_reduction.py (hunk 11 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_reduction.py:63]: You are absolutely correct, fixed

File: ngraph/python/tests/test_onnx/test_ops_matmul.py [hunk 12]
```diff
@@ -99,33 +106,20 @@ def import_and_compute_gemm(input_a, input_b, input_c, **kwargs):
...
-    assert np.array_equal(import_and_compute_matmul(*data), np.matmul(*data))
-
-    data = ([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10], [11, 12]])
-    assert np.array_equal(import_and_compute_matmul(*data), np.matmul(*data))
-
-    data = ([[1, 2], [3, 4], [5, 6]], [[7, 8, 9], [10, 11, 12]])
+@pytest.mark.parametrize(
+    "data, description",
+    [
+        pytest.param(([1, 2], [1, 3]), "vector at vector 1", marks=xfail_issue_35916),
```
Comment (postrational) [inline ngraph/python/tests/test_onnx/test_ops_matmul.py:112]: `@` means the matmul operation here, not `at`
```suggestion
        pytest.param(([1, 2], [1, 3]), "vector @ vector 1", marks=xfail_issue_35916),
```

File: ngraph/python/tests/test_onnx/test_ops_reduction.py [hunk 13]
```diff
@@ -17,7 +17,33 @@
 import onnx
 import pytest
 
-from tests.test_onnx.utils import run_node
+from tests.test_onnx.utils import run_node, unstrict_xfail_issue_35925, strict_xfail_issue_35925
+
+reduce_max = "ReduceMax"
```
Comment (postrational) [inline ngraph/python/tests/test_onnx/test_ops_reduction.py:22]: If these constants are used only once in the entire codebase, let's remove them and use strings directly instead.

File: ngraph/python/tox.ini (hunk 4 above)
Comment (postrational) [inline ngraph/python/tox.ini:26]:
```suggestion
  pytest --backend={env:NGRAPH_BACKEND} tests/test_ngraph -v
```

Comment (rblaczkowski) [reply ngraph/python/tox.ini:26]: It will be enabled in the new PR

File: ngraph/python/tests/test_onnx/test_ops_reduction.py (hunk 13 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_reduction.py:22]: OK, fixed

File: ngraph/python/tests/test_onnx/test_ops_matmul.py (hunk 12 above)
Comment (rblaczkowski) [reply ngraph/python/tests/test_onnx/test_ops_matmul.py]: changed to "and"
